src/V3Const.cpp

```diff
@@ -3348,7 +3340,7 @@ class ConstVisitor final : public VNVisitor {
     TREEOPA("AstNodeQuadop{$lhsp.castConst, $rhsp.castConst, $thsp.castConst, $fhsp.castConst}",  "replaceConst(nodep)");
     // Zero on one side or the other
     TREEOP ("AstAdd   {$lhsp.isZero, $rhsp}",   "replaceWRhs(nodep)");
-    TREEOP ("AstAnd   {$lhsp.isZero, $rhsp, isTPure($rhsp)}",   "replaceZero(nodep)");  // Can't use replaceZeroChkPure as we make this pattern in ChkPure
+    TREEOP ("AstAnd   {$lhsp.isZero, $rhsp, $rhsp.isPure}",   "replaceZero(nodep)");  // Can't use replaceZeroChkPure as we make this pattern in ChkPure
```

Comment:
I think we can optimize this - please update tests to make sure this is covered.

```suggestion
    TREEOP ("AstAnd   {$lhsp.isZero, $rhsp, $rhsp.isPure}",   "replaceZero(nodep)");  // Can't use replaceZeroChkPure as we make this pattern in ChkPure
    TREEOP ("AstAnd   {$lhsp.isZero, $rhsp, !$rhsp.isPure}",   "replaceWRhs(nodep)");  // Keep side-effect
```

Reply:
I think it's incorrect. This way we change the value of the expression. In case when LHS is 0 and RHS has side effects, the RHS should be evaluated, but the value of expression should be 0.

Reply:
Right.